packages/react/src/components/CodeSnippet/CodeSnippet.js

```diff
@@ -6,223 +6,170 @@
  */
 
 import PropTypes from 'prop-types';
-import React, { Component } from 'react';
+import React, { useState, useRef, useLayoutEffect } from 'react';
 import classNames from 'classnames';
 import { ChevronDown16 } from '@carbon/icons-react';
 import { settings } from 'carbon-components';
 import Copy from '../Copy';
+import Button from '../Button';
 import CopyButton from '../CopyButton';
-import uid from '../../tools/uniqueId';
+import getUniqueId from '../../tools/uniqueId';
 
 const { prefix } = settings;
 
-export default class CodeSnippet extends Component {
-  static propTypes = {
-    /**
-     * Provide the type of Code Snippet
-     */
-    type: PropTypes.oneOf(['single', 'inline', 'multi']),
-
-    /**
-     * Specify an optional className to be applied to the container node
-     */
-    className: PropTypes.string,
-
-    /**
-     * Provide the content of your CodeSnippet as a string
-     */
-    children: PropTypes.string,
-
-    /**
-     * Specify the string displayed when the snippet is copied
-     */
-    feedback: PropTypes.string,
-
-    /**
-     * Specify the description for the Copy Button
-     */
-    copyButtonDescription: PropTypes.string,
-
-    /**
-     * An optional handler to listen to the `onClick` even fired by the Copy
-     * Button
-     */
-    onClick: PropTypes.func,
-
-    /**
-     * Specify a label to be read by screen readers on the containing <textbox>
-     * node
-     */
-    copyLabel: PropTypes.string,
-
-    /**
-     * Specify a label to be read by screen readers on the containing <textbox>
-     * node
-     */
-    ariaLabel: PropTypes.string,
-
-    /**
-     * Specify a string that is displayed when the Code Snippet text is more
-     * than 15 lines
-     */
-    showMoreText: PropTypes.string,
-
-    /**
-     * Specify a string that is displayed when the Code Snippet has been
-     * interacted with to show more lines
-     */
-    showLessText: PropTypes.string,
-
-    /**
-     * Specify whether you are using the light variant of the Code Snippet,
-     * typically used for inline snippet to display an alternate color
-     */
-    light: PropTypes.bool,
-  };
-
-  static defaultProps = {
-    type: 'single',
-    showMoreText: 'Show more',
-    showLessText: 'Show less',
-  };
-
-  state = {
-    shouldShowMoreLessBtn: false,
-    expandedCode: false,
-  };
-
-  componentDidMount() {
-    if (this.codeContent) {
-      if (this.codeContent.getBoundingClientRect().height > 255) {
-        this.setState({ shouldShowMoreLessBtn: true });
-      }
+function CodeSnippet({
+  className,
+  type,
+  children,
+  feedback,
+  onClick,
+  ariaLabel,
+  copyLabel, //TODO: Merge this prop to `ariaLabel` in `v11`
+  copyButtonDescription,
+  light,
+  showMoreText,
+  showLessText,
+  ...other
+}) {
+  const [expandedCode, setExpandedCode] = useState(false);
+  const [shouldShowMoreLessBtn, setShouldShowMoreLessBtn] = useState(false);
+  const { current: uid } = useRef(getUniqueId());
```

Comment:
One thing that I don't think we remember is that technically we should not have ids start with a number, I think it's fine in HTML5 but in general can cause issues in certain cases (or at least has been brought up that way). Would we want to bring over our `useId` hook or add a prefix to this that's been passed to `id`?

Reply:
@joshblack our uniqueId actually adds an `id` prefix by default, so starting with a number shouldn't be an issue. Do you think we should be specifying a more descriptive prefix? I'm a little gun shy about tweaking ID generation behavior at this point haha.

Reply:
Ah, didn't realize it was using the older file for unique ids. In general, we use https://github.com/carbon-design-system/carbon/blob/master/packages/react/src/tools/setupGetInstanceId.js for newer code

Reply:
@joshblack oh missed that. I think this is a great opportunity to switch to a hook. I really highly doubt someone's depending on the ID of the inline code component, but do you think it'd be worth keeping the `id` prefix?

Reply:
It's not that an external user is leveraging it, just that folks have pointed out that ids generated with a number as the first character technically doesn't adhere to the HTML4 spec (I think this is addressed in HTML5). I think we could totally do `id-{number}`, we just have to be careful around collisions and the reason this came up with the newer approach is that it's easy to collide since they don't all share the same global id variable (which helps out a ton for tests). I think combining the local prefix plus the global id generation in a hook would be great and `useId` could help do that

Reply:
I was trying to avoid altering any external/dom behavior changes in this PR. I think if we want to change the unique id generation, we might want to do do it all at once, since it could potentially break folks using the id's for testing.

Reply:
I opened https://github.com/carbon-design-system/carbon/pull/4676 since I feel like changing the id generation behavior might be a little beyond the scope. The hook we have in the hooks package forces a `-` that doesn't exist in the current uniqueId implementation so it'd break some snapshots probably.

Reply:
@vpicone are we or others using automatically generated ids in tests?

Reply:
@joshblack wasn’t that the impetus for that big slack thread? Someone was using a generated ID to select an element in testing.

Reply:
@vpicone that was when someone used an `id` they passed into the component, not a generated one.

Reply:
Oh!!! That makes way more sense.